### PR TITLE
Enable speaker sounds

### LINKED ISSUE #298: Implement speaker sounds
Now that we have TIM1 available.

### PR DESCRIPTION
**Pull request by Peque**
Just for fun! :smile:

Fixes #298, #176. Closes #313.

### CODE REVIEW

File: src/speaker.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#include "speaker.h"
+
+uint8_t BASE_NOTE = 9;
+uint8_t BASE_OCTAVE = 4;
+float BASE_FREQUENCY = 440.;
+uint8_t POSITIONS[8] = {9, 11, 0, 2, 4, 5, 7};
```
Comment (cua-cua) [inline src/speaker.c]: Explain these numbers

File: src/main.c [hunk 2]
```diff
@@ -46,6 +47,24 @@ static void user_configuration(bool run)
...
+
+	voltage = get_battery_voltage();
+	if (voltage < 3.6)
+		speaker_warn_low_battery();
+	if (voltage < 3.5)
+		speaker_warn_low_battery();
+	if (voltage < 3.4)
+		speaker_warn_low_battery();
+	if (voltage < 3.3)
+		speaker_play('C', 3, 0, 2.);
```
Comment (cua-cua) [inline src/main.c]: I would create a function on hmi for this case, so speaker is only connected to hmi and is clearer

File: src/main.c [hunk 3]
```diff
@@ -46,6 +47,24 @@ static void user_configuration(bool run)
...
+/**
+ * @brief Check battery voltage and warn if the voltage is getting too low.
+ */
+static void check_battery_voltage(void)
+{
+	float voltage;
+
+	voltage = get_battery_voltage();
+	if (voltage < 3.6)
+		speaker_warn_low_battery();
```
Comment (cua-cua) [inline src/main.c]: Why not only one if with conditions:  (voltage < 3.6) && (voltage > 3.2)? why not else if?

File: src/speaker.c [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * Even if this function uses the scientific notation, notes are played with
+ * the concert pitch (standard pitch). That means A above middle C is the
+ * reference note, played at 440 Hz.
+ *
+ * @param[in] note Which note to play, in scientific notation.
+ * @param[in] octave Which octave to play, in scientific notation.
+ * @param[in] accidental Number of semitones to sum to the note.
+ * @param[in] duration Duration of the note, in seconds.
+ */
```
Comment (cua-cua) [inline src/speaker.c]: Include a @note, explaining that the the speaker and the emitters are sharing the timer 1 so they can not work simultaneously.

File: src/speaker.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	int16_t sound;
+	float frequency;
+
+	sound = POSITIONS[note - 'A'] - BASE_NOTE + (octave - BASE_OCTAVE) * 12;
+	sound += accidental;
+	frequency = pow(2, sound / 12.) * BASE_FREQUENCY;
+	setup_speaker();
+	speaker_set_frequency(frequency);
+	speaker_on();
+	sleep_ticks((uint32_t)(duration * 1000));
```
Comment (cua-cua) [inline src/speaker.c]: macro for 1000? S_TO_MS?

File: src/main.c (hunk 3 above)
Comment (Peque) [reply src/main.c]: To chain "low battery" sounds. Would give an idea on how low the battery is.

- 2 sounds for < 3.6
- 4 sounds for < 3.5
- ...
- Many sounds for < 3.3 plus a long sound for "critically low".

File: src/speaker.c (hunk 1 above)
Comment (Peque) [reply src/speaker.c]: Fixed in 9c06a76717413974d25939122a23e8d7601f7cd4.

File: src/speaker.c (hunk 5 above)
Comment (Peque) [reply src/speaker.c]: Fixed in 0a40b68ffc62b58b1d5581440e182415ed45b260

File: src/speaker.c (hunk 4 above)
Comment (Peque) [reply src/speaker.c]: Fixed in 6383d53035b08cf26c979dcda0ab48d79f3d0303

File: src/speaker.c (hunk 5 above)
Comment (Peque) [reply src/speaker.c]: Had to fixup that commit with c445b410284a9035e7facb745b608065266be168

Apparently, we cannot implement a `void sleep(float seconds)` function as `unistd.h` is already defining a sleep function (that is implemented with an `unsigned int` parameter instead). :confused:

File: src/speaker.c (hunk 1 above)
Comment (cua-cua) [reply src/speaker.c]: I like ti <3

File: src/speaker.c (hunk 5 above)
Comment (cua-cua) [reply src/speaker.c]: Very elegant ;)

**Pull request merged**